flutter/lib/ui/geometry.dart

```diff
@@ -680,35 +680,183 @@ class Rect {
+  @override
+  int get hashCode => hashValues(x, y);
+
+  @override
+  String toString() {
+    return x == y ? 'Radius.circular($x)' : 'Radius.oval($x, $y)';
+  }
+}
+
+/// A rounded rectangle with the custom radii for all four corners.
 class RRect {
   RRect._();
 
   /// Construct a rounded rectangle from its left, top, right, and bottom edges,
-  /// and the radii along its horizontal axis and its vertical axis.
-  RRect.fromLTRBXY(double left, double top, double right, double bottom, double radiusX, double radiusY) {
+  /// and the constant radii along its horizontal axis and its vertical axis.
+  RRect.fromLTRBXY(double left, double top, double right, double bottom,
+                   double radiusX, double radiusY) {
     _value
       ..[0] = left
       ..[1] = top
       ..[2] = right
       ..[3] = bottom
       ..[4] = radiusX
-      ..[5] = radiusY;
+      ..[5] = radiusY
+      ..[6] = radiusX
+      ..[7] = radiusY
+      ..[8] = radiusX
+      ..[9] = radiusY
+      ..[10] = radiusX
+      ..[11] = radiusY;
+  }
+
+  /// Construct a rounded rectangle from its left, top, right, and bottom edges,
+  /// and a radius that is constant in each corner.
+  RRect.fromLTRBR(double left, double top, double right, double bottom,
+                      Radius radius) {
+    _value
+      ..[0] = left
+      ..[1] = top
+      ..[2] = right
+      ..[3] = bottom
+      ..[4] = radius.x
+      ..[5] = radius.y
+      ..[6] = radius.x
+      ..[7] = radius.y
+      ..[8] = radius.x
+      ..[9] = radius.y
+      ..[10] = radius.x
+      ..[11] = radius.y;
   }
 
-  /// Construct a rounded rectangle from its bounding box and the radii along
-  /// its horizontal axis and its vertical axis.
+  /// Construct a rounded rectangle from its bounding box and the constant radii
+  /// along its horizontal axis and its vertical axis.
   RRect.fromRectXY(Rect rect, double radiusX, double radiusY) {
     _value
       ..[0] = rect.left
       ..[1] = rect.top
       ..[2] = rect.right
       ..[3] = rect.bottom
       ..[4] = radiusX
-      ..[5] = radiusY;
+      ..[5] = radiusY
+      ..[6] = radiusX
+      ..[7] = radiusY
+      ..[8] = radiusX
+      ..[9] = radiusY
+      ..[10] = radiusX
+      ..[11] = radiusY;
   }
 
-  static const int _kDataSize = 6;
+  /// Construct a rounded rectangle from its bounding box and a radius that is
+  /// constant in each corner.
+  RRect.fromRectAndRadius(Rect rect, Radius radius) {
+    _value
+      ..[0] = rect.left
+      ..[1] = rect.top
+      ..[2] = rect.right
+      ..[3] = rect.bottom
+      ..[4] = radius.x
+      ..[5] = radius.y
+      ..[6] = radius.x
+      ..[7] = radius.y
+      ..[8] = radius.x
+      ..[9] = radius.y
+      ..[10] = radius.x
+      ..[11] = radius.y;
+  }
+
+  /// Construct a rounded rectangle from its left, top, right, and bottom edges,
+  /// and topLeft, topRight, bottomRight, and bottomLeft radii.
+  RRect.fromLTRBCustom(double left, double top, double right, double bottom,
```

Comment:
I don't like "Custom" in the constructor name here, but I don't have a better idea. @abarth, suggestions?